File: algorithms/112.swift
// https://leetcode.com/problems/path-sum

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func hasPathSum(root: TreeNode?, _ sum: Int) -> Bool {
        func dfs_(node: TreeNode?, _ currentSum: Int, _ targetSum: Int) -> Bool {
            guard let n = node else { return false }
            let c = currentSum + n.val
            if n.left == nil && n.right == nil { return c == targetSum }
            else { return dfs_(n.left, c, targetSum) || dfs_(n.right, c, targetSum) }
        }
        return dfs_(root, 0, sum)
    }
}

File: algorithms/112.py
# https://leetcode.com/problems/path-sum


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """

        def dfs_(node, currentSum, targetSum):
            if node is None:
                return False

            currentSum += node.val
            if node.left is None and node.right is None:
                return currentSum == targetSum
            return dfs_(node.left, currentSum, targetSum) or dfs_(node.right, currentSum, targetSum)

        return dfs_(root, 0, sum)
